vpc/globalgc/globalgc.go

```diff
@@ -4,303 +4,175 @@ import (
+		if err != nil {
+			return err
+		}
+		for _, networkInterface := range describeAvailableResult.NetworkInterfaces {
+			networkInterfaceChannel <- networkInterface
+		}
+		if describeAvailableResult.NextToken == nil {
+			return nil
+		}
+		describeAvailableRequest.SetNextToken(*describeAvailableResult.NextToken)
 	}
-	return nil
 }
 
-func doZonalGC(parentCtx *context.VPCContext, minDetachTime time.Duration, az string, vpcID string) error {
-	ec2Client := ec2.New(parentCtx.AWSSession)
-
-	baseFilters := getBaseFilter(vpcID, az)
-	parentCtx.Logger.WithField("filters", baseFilters).Debug("Configuring ENI fetch filters")
-
-	describeAvailableRequest := &ec2.DescribeNetworkInterfacesInput{
-		Filters: append([]*ec2.Filter{
-			{
-				Name:   aws.String("status"),
-				Values: aws.StringSlice([]string{"available"}),
-			},
-		}, baseFilters...),
+// TODO: Find fatal errors and return errors
+func cleanupENI(parentCtx *context.VPCContext, svc *ec2.EC2, detachTime time.Duration, eni *ec2.NetworkInterface) error {
+	ctx := parentCtx.WithField("NetworkInterfaceId", *eni.NetworkInterfaceId)
+	tags := ec2util.TagSetToMap(eni.TagSet)
+	networkCreationTime, ok := tags[vpc.ENICreationTimeTag]
+	if !ok {
+		ctx.Logger.Warning("ENI does not have creation time field")
+		return nil
 	}
-
-	networkInterfaces, err := ec2Client.DescribeNetworkInterfacesWithContext(parentCtx, describeAvailableRequest)
+	creationTime, err := time.Parse(time.RFC3339, *networkCreationTime)
 	if err != nil {
-		return err
+		ctx.Logger.WithField(vpc.ENICreationTimeTag, *networkCreationTime).WithError(err).Error("Cannot parse")
+		return nil
 	}
-
-	now := time.Now()
-	// Get the candidates, and selected
-	candidates, selected := markAndCollect(parentCtx, networkInterfaces, minDetachTime, now)
-	parentCtx.Logger.Info("Going to GC: ", selected)
-	// Delete the selected
-	err = deleteSelected(parentCtx, selected)
-	if err != nil {
-		return err
+	timeSinceCreation := time.Since(creationTime)
+	if timeSinceCreation < detachTime {
+		ctx.Logger.WithField("timeSinceCreation", timeSinceCreation.String()).Debug("Not cleaning up ENI, too young")
+		return nil
 	}
 
-	// Mark the candidates
-	if len(candidates) > 0 {
-		err = doMark(parentCtx, candidates, now, ec2Client)
-		if err != nil {
-			return err
-		}
+	if *eni.Description != vpc.NetworkInterfaceDescription {
+		panic(fmt.Sprintf("Interface description is %s instead of %s", *eni.Description, vpc.NetworkInterfaceDescription))
 	}
-
-	// Remove incorrectly marked interfaces
-	describeMarkedRequest := &ec2.DescribeNetworkInterfacesInput{
-		Filters: append([]*ec2.Filter{
-			{
-				Name:   aws.String("status"),
-				Values: aws.StringSlice([]string{"in-use"}),
-			},
-			{
-				Name:   aws.String("tag-key"),
-				Values: aws.StringSlice([]string{markTag}),
-			},
-		}, baseFilters...),
+	if *eni.Status != "available" {
+		panic(fmt.Sprintf("Interface status is %s instead of available", *eni.Status))
```

Comment:
shouldn't this just skip this interface? (or is this the TODO on this function?)